Optimize rate calculator (v2)

vpalmisano opened the pull request:
With this PR we are fixing issues generated by the new RateCalculator implementation:
- it uses a window granularity `windowSize / windowItems` to sum up multiple samples into the newest item;
- in the default configuration we are using a total items count of `100`, this means that with a window size of `1000ms` each item will sum up the samples arriving in a `10ms`  time window;
- avoids to fill the calculation buffer;
- fixed oldestTime/Index naming

vpalmisano committed:
rate calculator optimization

vpalmisano committed:
applied format:worker

vpalmisano committed:
comment format

vpalmisano committed:
comment format

vpalmisano committed:
comment format

vpalmisano committed:
renamed latestTime, latestIndex and oldestIndex

vpalmisano committed:
added a warning log when the calculation buffer is full

vpalmisano committed:
Merge branch 'optimize-rate-calculator' of https://github.com/vpalmisano/mediasoup into optimize-rate-calculator

vpalmisano committed:
Update worker/src/RTC/RateCalculator.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

vpalmisano committed:
Update worker/src/RTC/RateCalculator.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

vpalmisano committed:
Merge branch 'v3' into optimize-rate-calculator

vpalmisano committed:
using the window granularity to sum up multiple samples into the newest item; avoids to fill the calculation buffer; fixed oldestTime/Index naming

vpalmisano commented on worker/src/RTC/RateCalculator.cpp:
```diff
@@ -12,46 +12,57 @@ namespace RTC
+				this->newestTimeIndex = 0;
+
+			// Newest index overlaps with the oldest one, remove it.
+			if (this->newestTimeIndex == this->oldestTimeIndex && this->oldestTimeIndex != -1)
```
This condition should never happen with the new implementation.

vpalmisano commented on worker/test/src/RTC/TestRateCalculator.cpp:
```diff
@@ -61,7 +61,7 @@ SCENARIO("Bitrate calculator", "[rtp][bitrate]")
 	SECTION("slide")
 	{
-		RateCalculator rate;
+		RateCalculator rate(1000, 8000, 1000);
```
With the default window items value (`100`) we have this:
```
		RateCalculator rate(1000, 8000, 100);

		std::vector<data> input =
		{
			{ 0,    5, 40 },
			{ 999,  2, 56 },
			{ 1001, 1, 24 }, // merged inside 999
			{ 1001, 1, 32 }, // merged inside 999
			{ 2000, 1, 24 } // will erase the item with timestamp=999, removing also the next two samples. The end estimation will include only the last sample 
		};
```

vpalmisano committed:
added more comments to the slide test

penguinol commented on worker/src/RTC/RateCalculator.cpp:
```diff
@@ -12,46 +12,57 @@ namespace RTC
-		if (this->newestTimeIndex == this->oldestIndex && this->oldestIndex != -1)
+		// If the elapsed time from the last used item is greater than the
+		// window granularity, increase the index.
+		if (this->newestTimeIndex < 0 || nowMs - this->newestTime >= windowSize / windowItems)
```
maybe we can save the result of windowSize/windowItem to avoid dividing it every time

jmillan replied:
Yes, it indicates the milliseconds represented in each item (ie: `this->itemSize = this->windowSize/this->windowItems`). It should be a class member IMHO.

vpalmisano replied:
Added with the last commits.

jmillan commented on worker/include/RTC/RateCalculator.hpp:
```diff
@@ -46,7 +46,7 @@ namespace RTC
 			this->newestTimeIndex = -1;
 			this->oldestTimeIndex = 0u;
-			this->oldestIndex     = -1;
+			this->oldestTimeIndex = -1;
```
This and previous line set the same member.

vpalmisano replied:
Fixed.

jmillan commented:
I'll properly review this tomorrow.

vpalmisano committed:
fixed reset function

vpalmisano committed:
store windowSize / windowItems into the itemSize property

vpalmisano committed:
added a min value constraint

jmillan commented on worker/src/RTC/RateCalculator.cpp:
```diff
@@ -12,46 +12,57 @@ namespace RTC
-		// Newest index overlaps with the oldest one, remove it.
-		if (this->newestTimeIndex == this->oldestIndex && this->oldestIndex != -1)
+		// If the elapsed time from the last used item is greater than the
+		// window granularity, increase the index.
```
The condition below is not accurate.

Each item in the buffer must contain the data for a certain period of time. Ie:

Scenario where each item holds 10ms of time.

`windowSize = 1000`
`windowItems = 100`
`oldestTime = 2000`

|Index|0    |1       |2        |3          | n|        
|------|----|-----|------|-------|--|
| Time Range|2000-2009|2010-2019|2020-2029|2030-2039|.. |

So, instead of having `newestTime`, we should have the start time for the newest item, and check whether the incoming data should be added to the same item or not.

vpalmisano replied:
It is already there, the newestTime update is performed only when a new item is added: https://github.com/vpalmisano/mediasoup/blob/optimize-rate-calculator/worker/src/RTC/RateCalculator.cpp#L28

jmillan replied:
Right.

I'm having a hard time reading the PR due to the member naming. I propose some naming enhancement:

newestTimeIndex -> newestItemIndex (Buffer index holding the newest item)
oldestTimeIndex -> oldestItemIndex (Buffer index holding the oldest item)
newestTime -> newestItemStartTime (Start time of the newest item)
oldestTime -> oldestItemStartTime (Start time of the oldest item)
windowSize -> windowSizeMs (Window size in milliseconds)
itemSize -> itemSizeMs (Item size in milliseconds)

It would certainly help with the code readability.

What do you think?

vpalmisano replied:
I agree, much better. Done.

vpalmisano committed:
renaming vars

jmillan commented on worker/src/RTC/RateCalculator.cpp:
```diff
@@ -12,46 +12,57 @@ namespace RTC
-
-		// Newest index overlaps with the oldest one, remove it.
-		if (this->newestTimeIndex == this->oldestIndex && this->oldestIndex != -1)
+		// If the elapsed time from the last used item is greater than the
+		// window granularity, increase the index.
```
```suggestion
		// If the elapsed time from the newest item start time is greater than the
		// item size (in milliseconds), increase the item index.
```

jmillan commented:
@vpalmisano,

The PR looks good to me. The new variable names makes it much more readable, thanks!

Just one thing. Please, instead of modifying the existing tests, keep them, and create new ones with the new `windowSizeMs` and `windowItems` values.

vpalmisano committed:
Update worker/src/RTC/RateCalculator.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

vpalmisano commented:
> Just one thing. Please, instead of modifying the existing tests, keep them, and create new ones with the new `windowSizeMs` and `windowItems` values.

The only modification I've done is here (https://github.com/vpalmisano/mediasoup/blob/optimize-rate-calculator/worker/test/src/RTC/TestRateCalculator.cpp#L73) and it is required to make the test pass as explained in the comment.

jmillan commented:
What I mean is, keep those two tests with the original construction values `(1000, 8000, 1000)` and a create two new ones with the new values.

The old tests should keep passing providing the correct values in construction.

<img width="1646" alt="Screenshot 2021-04-29 at 09 33 30" src="https://user-images.githubusercontent.com/732340/116516643-21479200-a8ce-11eb-84d6-7a18faeb8a4b.png">

vpalmisano committed:
reverted slide test to previous version and added new slide test with a 100 items window

vpalmisano commented:
> What I mean is, keep those two tests with the original construction values `(1000, 8000, 1000)` and a create two new ones with the new values.
> 
> The old tests should keep passing providing the correct values in construction.

Yes, the "slide" test will pass using `(1000, 8000, 1000)`, I've added a new one with `(1000, 8000, 100)` values.

The "wrap" test is useful only if new samples are allocated to different items, for this reason I've modified the samples timestamps.

jmillan commented:
@vpalmisano,

This warning: https://github.com/vpalmisano/mediasoup/blob/ff423f531b9a526ef45f71e72c7255d21aa402af/worker/src/RTC/RateCalculator.cpp#L35

Shouldn't be so. It's a legit situation. Any reason why you set a warning there?

vpalmisano commented:
> @vpalmisano,
> 
> This warning: https://github.com/vpalmisano/mediasoup/blob/ff423f531b9a526ef45f71e72c7255d21aa402af/worker/src/RTC/RateCalculator.cpp#L35
> 
> Shouldn't be so. It's a legit situation. Any reason why you set a warning there?

The if condition should never happen with the latest modification. 
I left the warning in the case we have missed something.
